File: 1-two-sum/1-two-sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        """
        declare hashtable
          (this will store K:V as value:idx)
        iterate through nums
          if target - curr num is in hashset, return [hashtable[val], curr i]
          else, store the num in the hashtable with the idx as the value
        """
        comps = {}
        for i, num in enumerate(nums):
          if target - num in comps:
            return [comps[target - num], i]
          else:
            comps[num] = i

File: 1-two-sum/1-two-sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        // return bruteForce(nums, target);
        return map(nums, target);
    }
    
    public int[] map(int[] nums, int target) {
        // is the compliment to curr in there? if so, return. else, just put this num and the index because it could be someone elses compliment
        Map<Integer, Integer> compliments = new HashMap<>();
        
        for (int i = 0; i < nums.length; ++i) {
            int compliment = target - nums[i];
            
            if (compliments.containsKey(compliment)) {
                return new int[] { i, compliments.get(compliment) };
            }
            
            compliments.put(nums[i], i);
        }
        
        return null;
    }
    
    public int[] bruteForce(int[] nums, int target) {
        for (int i = 0; i < nums.length; ++i) {
            for (int j = i + 1; j < nums.length; ++j) {
                if (nums[i] + nums[j] == target) {
                    return new int[] { i, j };
                }
            }
        }
        
        return null;
    }
}